### PR TITLE
cache buffer loads across multiple bufs

### PR DESCRIPTION
**Pull request by chaosagent**
For winograd #1037; follow up to #1272. Different bufs can load the same stuff from the same rawbufs. The use case is Tensor.stack with the stack dimension unrolled.

### CODE REVIEW

**Comment by tinyb0t** [comment]
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/codegen/linearizer.py            10               7
------------------------------------------------------------
lines added in the tinygrad folder: 3
```

**Comment by geohot** [comment]
These PRs are really hard to follow. Can you show an example of what it changes?

**Comment by geohot** [comment]
I think my ask was clear.

**Comment by chaosagent** [comment]
Before (just running the added Linearizer test):

```
*** exec  0.00 GB    0.02 ms op: BinaryOps.ADD        out(float): (3,)                           in(2): [(3,)] 
UOps.DEFINE_GLOBAL  :                           []                               ('data0', dtypes.float)
UOps.DEFINE_GLOBAL  :                           []                               ('data1', dtypes.float)
   0 buffer<3, dtypes.float>                         [View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True, shape_strides=((3, 1),))]
   1 buffer<4, dtypes.float>                         [View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True, shape_strides=((3, 1),))]
   2 buffer<4, dtypes.float>                         [View(shape=(3,), strides=(1,), offset=1, mask=None, contiguous=False, shape_strides=((3, 1),))]
   3
UOps.LOOP           :                           []                               ([], 'global')
UOps.LOOP           :                           []                               ([], 'local')
UOps.LOAD           : <val1_0>                  []                               MemOp(name='data1', idx=<0>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val1_1>                  []                               MemOp(name='data1', idx=<1>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val1_2>                  []                               MemOp(name='data1', idx=<2>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val2_0>                  []                               MemOp(name='data1', idx=<1>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val2_1>                  []                               MemOp(name='data1', idx=<2>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val2_2>                  []                               MemOp(name='data1', idx=<3>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.ALU            : <alu0>                    [<val1_0>, <val2_0>]             BinaryOps.ADD
UOps.ALU            : <alu1>                    [<val1_1>, <val2_1>]             BinaryOps.ADD
UOps.ALU            : <alu2>                    [<val1_2>, <val2_2>]             BinaryOps.ADD
UOps.STORE          :                           [<alu0>]                         MemOp(name='data0', idx=<0>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.STORE          :                           [<alu1>]                         MemOp(name='data0', idx=<1>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.STORE          :                           [<alu2>]                         MemOp(name='data0', idx=<2>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.ENDLOOP        :                           []                               ([], 'global+local')

```


After:

```
*** exec  0.00 GB    0.02 ms op: BinaryOps.ADD        out(float): (3,)                           in(2): [(3,)] 
UOps.DEFINE_GLOBAL  :                           []                               ('data0', dtypes.float)
UOps.DEFINE_GLOBAL  :                           []                               ('data1', dtypes.float)
   0 buffer<3, dtypes.float>                         [View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True, shape_strides=((3, 1),))]
   1 buffer<4, dtypes.float>                         [View(shape=(3,), strides=(1,), offset=0, mask=None, contiguous=True, shape_strides=((3, 1),))]
   2 buffer<4, dtypes.float>                         [View(shape=(3,), strides=(1,), offset=1, mask=None, contiguous=False, shape_strides=((3, 1),))]
   3
UOps.LOOP           :                           []                               ([], 'global')
UOps.LOOP           :                           []                               ([], 'local')
UOps.LOAD           : <val1_0>                  []                               MemOp(name='data1', idx=<0>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val1_1>                  []                               MemOp(name='data1', idx=<1>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val1_2>                  []                               MemOp(name='data1', idx=<2>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.LOAD           : <val2_3>                  []                               MemOp(name='data1', idx=<3>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.ALU            : <alu0>                    [<val1_0>, <val1_1>]             BinaryOps.ADD
UOps.ALU            : <alu1>                    [<val1_1>, <val1_2>]             BinaryOps.ADD
UOps.ALU            : <alu2>                    [<val1_2>, <val2_3>]             BinaryOps.ADD
UOps.STORE          :                           [<alu0>]                         MemOp(name='data0', idx=<0>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.STORE          :                           [<alu1>]                         MemOp(name='data0', idx=<1>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.STORE          :                           [<alu2>]                         MemOp(name='data0', idx=<2>, local=False, memory_dtype=dtypes.float, valid=<1>, invalid_value=0.0)
UOps.ENDLOOP        :                           []                               ([], 'global+local')


```
